registry/storage/driver/speedy/speedyclient.go

```diff
@@ -0,0 +1,378 @@
+package speedy
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+)
+
+//Client is the client for speedy.
+type Client struct {
+}
+
+//MetaInfoValue is meta info format of speedy.
+type MetaInfoValue struct {
+	Index   uint64
+	Start   uint64
+	End     uint64
+	IsLast  bool
+	IsDir   bool `json:",omitempty"`
+	ModTime time.Time
+}
+
+//OrderByIndex is Used to sort metainfo by index
+//E.g. we may get a metainfo array for a special path, so we need to sort this metainfo array.
+type OrderByIndex []*MetaInfoValue
+
+const (
+	headerSourcePath = "Source-Path"
+	headerDestPath   = "Dest-Path"
+	headerPath       = "Path"
+	headerIndex      = "Fragment-Index"
+	headerRange      = "Bytes-Range"
+	headerIsLast     = "Is-Last"
+	headerVersion    = "Registry-Version"
+	fragmentInfo     = "fragment-info"
+	fileList         = "file-list"
+	pathDescendant   = "path-descendant"
+)
+
+func (a OrderByIndex) Len() int {
+	return len(a)
+}
+
+func (a OrderByIndex) Swap(i, j int) {
+	a[i], a[j] = a[j], a[i]
+}
+
+func (a OrderByIndex) Less(i, j int) bool {
+	return a[i].Index < a[j].Index
+}
+
+func (c *Client) doRequest(req *http.Request) ([]byte, int, error) {
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		if resp != nil {
+			return nil, resp.StatusCode, err
+		}
+		return nil, http.StatusNotFound, err
+	}
+
+	dataBody, err := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if err != nil {
+		return nil, resp.StatusCode, err
+	}
+	return dataBody, resp.StatusCode, nil
+}
+
+func (c *Client) getMetaInfoValueFromJSON(data []byte) ([]*MetaInfoValue, error) {
+	var mapResult map[string]interface{}
+	err := json.Unmarshal(data, &mapResult)
+	if err != nil {
+		return nil, err
+	}
+
+	fragmentInfoValue, ok := mapResult[fragmentInfo]
+	if !ok {
+		return nil, fmt.Errorf("Response format maybe error: %v", mapResult)
+	}
+
+	infoArr, ok := fragmentInfoValue.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("Response format maybe error, value is not a array: %v", fragmentInfoValue)
+	}
+
+	var result []*MetaInfoValue
+	for _, info := range infoArr {
+		m, ok := info.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("Response format maybe error, infoArr: %v", infoArr)
+		}
+
+		miv := new(MetaInfoValue)
+		index, ok := m["Index"].(float64)
+		if !ok {
+			return nil, fmt.Errorf("Response format maybe error, MetaInfoValue: %v", m)
+		}
+		miv.Index = uint64(index)
```

Comment:
index is uint64, 
but json.Marshal will change int(int64 and so on) to float64
so the int value in the result of json.Unmarshal is float64


for example:
package main


import (
"encoding/json"
"fmt"
)


type Test struct {
Index int
Start int
}


func main() {
t := &Test{
Index: int(10),
Start: int(11),
}
fmt.Println(t)


byteResult, err := json.Marshal(t)
if err != nil {
fmt.Println(err)
return
}


fmt.Println("result:", byteResult)


var m map[string]interface{}
err = json.Unmarshal(byteResult, &m)
if err != nil {
fmt.Println(err)
return
}
fmt.Println(m)


index, ok := m["Index"].(float64)
if !ok {
fmt.Println("the type of m[index] is not float64")
return
}
fmt.Println("the type of m[index] is float64")
fmt.Println(index)


index2, ok := m["Index"].(int)
if !ok {
fmt.Println("the type of m[index] is not int")
return
}
fmt.Println(index2)
}




--


At 2015-08-12 06:39:16, "Anton Tyurin" <notifications@github.com> wrote:


In registry/storage/driver/speedy/speedyclient.go:

> +		return nil, fmt.Errorf("Response format maybe error, value is not a array: %v", fragmentInfoValue)
> +	}
> +
> +	var result []*MetaInfoValue
> +	for _, info := range infoArr {
> +		m, ok := info.(map[string]interface{})
> +		if !ok {
> +			return nil, fmt.Errorf("Response format maybe error, infoArr: %v", infoArr)
> +		}
> +
> +		miv := new(MetaInfoValue)
> +		index, ok := m["Index"].(float64)
> +		if !ok {
> +			return nil, fmt.Errorf("Response format maybe error, MetaInfoValue: %v", m)
> +		}
> +		miv.Index = uint64(index)


Is index float64 or uint64? What's the point of casting to float64 & to uint64?

—
Reply to this email directly or view it on GitHub.